cylc/flow/etc/job.sh

```diff
@@ -193,6 +216,14 @@ cylc__job_finish_err() {
         wait "${CYLC_TASK_MESSAGE_STARTED_PID}" 2>'/dev/null' || true
     fi
     cylc message -- "${CYLC_SUITE_NAME}" "${CYLC_TASK_JOB}" "$@" || true
+    # Propagate real signals to entire process group, if we are a group leader,
+    # otherwise just to the backgrounded user script.
+    if [[ -n "${CYLC_TASK_SCRIPT_PID:-}" ]] &&
+       [[ ":DEBUG:ERR:EXIT:RETURN:" != *":${signal}:"* ]]; then
+        kill -s "${signal}" -- "-$$" 2>'/dev/null' ||
+        kill -s "${signal}" "${CYLC_TASK_SCRIPT_PID}" 2>'/dev/null' || true
+        wait  # in case child user script traps the signal
+    fi
```

Comment:
The logic here is to try to resend the signal to the entire process group (will succeed if jobscript shell is a group leader - should this be always the case?) or, if that fails, to the backgrounded user script subshell. User script is free set its own traps, including setting them to ignored for itself and/or its children. Then we wait for user script process to finish, so that there is no risk of running it concurrently with `err-script` (I think this is a sensible thing to do).

Reply:
Actually, there is problem with this. When bash receives a signal during `wait` command it returns from that command with exit code 128 + signal number. Because the error code is non-zero, ERR trap triggers before the signal trap and, due to `exit 1` in there, the signal trap never gets executed. This is pretty annoying, because distinguishing between signal and error exit becomes more complicated.

Reply:
The above is still not quite correct. It is what one might expect (at least what I would expect based on bash docs), but reality is more complicated. Just have a look at this simple test:
```
$ bash -c 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; f(){ sleep 5 & wait $!; }; f' & sleep 1; kill -- $!; wait
TERM:143
ERR:143
EXIT:143
$ bash -ec 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; f(){ sleep 5 & wait $!; }; f' & sleep 1; kill -- $!; wait
TERM:143
EXIT:143
$ bash -ec 'for sig in ERR EXIT TERM; do trap "echo $sig:\$?" $sig; done; sleep 5 & wait $!' & sleep 1; kill -- $!; wait
ERR:143
TERM:143
EXIT:143
$ bash --version
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
```
So the order of traps depends on whether or not the function is used. Moreover, doing the same with bash 3.2 I get TERM trap executed first in all these cases, so this behaviour is not even consistent bash between versions (at least the ones I have access to test with).

The conclusion from that is that one cannot trust the order of trap execution in bash. Special care is therefore needed to make sure that the right traps are triggered and in the right order. I think I have a working solution to this now and will push it up soon for discussion.

And now for the real kicker. While going down this rabbit hole I've come across this lovely post:
https://lists.gnu.org/archive/html/bug-bash/2012-12/msg00093.html

The gist of it is to never, ever use functions in conditionals in combination with `set -e` (unless you specifically design them for this) or you might be in for a really bad surprise (spoiler alert: POSIX mandates that `set -e` is simply disabled for anything executing within the conditional context with disastrous consequences for functions and compound commands). But read the linked post, it's worth it.

Reply:
I think if I had to choose between C, Fortran, and Shell, Shell would probably be my last option :eyes: Great detective work @TomekTrzeciak :clap:

Reply:
More of bash woes:
```
$ bash -ec 'f() { trap "echo xxx" EXIT; false; };  f & wait $!'; echo $?
xxx
1
$ bash -c 'f() { trap "echo xxx" EXIT; false; };  f & wait $!'; echo $?
1
$ bash -c 'f() { trap "echo xxx" EXIT; false; exit; };  f & wait $!'; echo $?
xxx
1
```
Looks like bash remembers to execute EXIT traps only some of the time. I get this with bash 4.2 and 3.2 alike.

Reply:
Even more bash insanity:
```
$ bash -c 'f() { for S in ERR EXIT; do trap "echo ${S}_CODE=\$?" ${S};  done; (exit 42); };  (f) & wait $!'
ERR_CODE=42
EXIT_CODE=42
$ bash -ec 'f() { for S in ERR EXIT; do trap "echo ${S}_CODE=\$?" ${S};  done; (exit 42); };  (f) & wait $!'
ERR_CODE=42
EXIT_CODE=1
$ bash --version
GNU bash, version 4.2.46(2)-release (x86_64-redhat-linux-gnu)
```
Somewhere between handling ERR and EXIT traps `bash -e` decides to change the status code! WTF?

(This works as expected on bash 3.2 or if you replace `(f) &` with `f &`, expect that the error trap is not executed then unless you also have `set -e`.)